src/ydata_synthetic/synthesizers/timeseries/tscwgan/model.py

```diff
@@ -0,0 +1,257 @@
+        g_loss = - reduce_mean(score_fake)
+        return g_loss
+
+    def _make_fake_batch(self, real_batch, noise_batch):
+        """Generate a batch of fake records and return it with the batch of used conditions.
+        Conditions are the first elements of records in the real batch."""
+        cond_batch = real_batch[:, :self.cond_dim]
+        gen_input = concat([cond_batch, noise_batch], axis=1)
+        return self.generator(gen_input, training=True), cond_batch
+
+    def gradient_penalty(self, real, fake):
+        gp = gradient_penalty(self.critic, real, fake, mode=Mode.DRAGAN)
+        return gp
+
+    def _generate_noise(self):
+        "Gaussian noise for the generator input."
+        while True:
+            yield normal(size=self.noise_dim)
+
+    def get_batch_noise(self):
+        "Create a batch iterator for the generator gaussian noise input."
+        return iter(tfdata.Dataset.from_generator(self._generate_noise, output_types=float32)
+                                .batch(self.batch_size)
+                                .repeat())
+
+    def get_batch_data(self, data, n_windows= None):
+        if not n_windows:
+            n_windows = len(data)
+        data = squeeze(convert_to_tensor(data, dtype=float32))
+        return iter(tfdata.Dataset.from_tensor_slices(data)
+                                .shuffle(buffer_size=n_windows)
+                                .batch(self.batch_size).repeat())
+
+    def sample(self, condition: array, n_samples: int = 100, seq_len: int = 24):
+        """For a given condition, produce n_samples of length seq_len.
+
+        Args:
+            condition (numpy.array): Condition for the generated samples, must have the same length.
+            n_samples (int): Minimum number of generated samples (returns always a multiple of batch_size).
+            seq_len (int): Length of the generated samples.
+
+        Returns:
+            data (numpy.array): An array of data of shape [n_samples, seq_len]"""
+        assert len(condition.shape) == 2, "Condition array should be two-dimensional."
+        assert condition.shape[1] == self.cond_dim, \
+            f"The condition sequence should have {self.cond_dim} length."
+        batches = n_samples // self.batch_size + 1
+        ar_steps = seq_len // self.data_dim + 1
+        data = []
+        z_dist = self.get_batch_noise()
+        for batch in trange(batches, desc=f'Synthetic data generation'):
+            data_ = []
+            cond_seq = convert_to_tensor(condition, float32)
+            gen_input = concat([tile(cond_seq, multiples=[self.batch_size, 1]), next(z_dist)], axis=1)
+            for step in range(ar_steps):
+                records = self.generator(gen_input, training=False)
+                gen_input = concat([records[:, -self.cond_dim:], next(z_dist)], axis=1)
+                data_.append(records)
+            data_ = hstack(data_)[:, :seq_len]
+            data.append(data_)
+        return array(vstack(data))
```

Comment:
Regarding to a revision of this sample method I think:

1. We can force the number of samples to be the same as requested, i.e. truncate the last batch if needed
2. Condition converted to tensor could be taken outside the for loop
3. It does not seem to be truly auto-regressive since only the condition part of the produced records is being used as condition, we can add this capacity

Do you agree/have anything to add @fabclmnt ?